[dagster-managed-elements] Getting started guide (#10224)
dagster-io/dagster: docs/content/guides/dagster/airbyte-managed-elements.mdx

---
title: Managed Airbyte instance | Dagster
---

# Mananged Airbyte instance <Experimental />

This guide provides an intro to using Dagster to configure your [Airbyte](/integrations/airbyte) connections. This allows you to centralize the configuration for your data stack, specifying configuration in Python code. You can check-in and version your config with version control or programatically generate config for more complex use-cases.

---

## Prerequisites

In order to use this feature, you must first install `dagster-airbyte` and the `dagster-managed-elements` package, which includes the base config reconcilation APIs and a CLI, available as `dagster-me`.

```bash
pip install dagster-airbyte dagster-managed-elements
```

## 1. Defining a reconciler

> The config for your Airbyte instance is specified in a `AirbyteManagedElementReconciler`, which is pointed at a specific Airbyte instance using an Airbyte resource. It is provided with a list of connections to reconcile, which we will set up in the following steps.

Reviewer: Thoughts on linking to the `Additional configuration options` section here?

> The config for your Airbyte instance is specified in a `AirbyteManagedElementReconciler`, which is pointed at a specific Airbyte instance using an Airbyte resource. It is provided with a list of connections to reconcile, which we will set up in the following steps.

Author: Provided a link out to the Airbyte guide which covers setting up the Resource in more detail.

    url="https://docs.dagster.io/assets/cereal.csv",
    format="csv",
    provider=FileSource.HTTPSPublicWeb(),
    dataset_name="cereals",
)

local_json_destination = LocalJsonDestination(
    name="local-json",
    destination_path="/local/cereals_out.json",
)
```

## 3. Defining a connection

Finally, we will define a connection between our source and destination. This is defined using the `AirbyteConnection` class.

```python startafter=start_define_connection endbefore=end_define_connection file=/guides/dagster/managed_elements/airbyte.py dedent=4
from dagster_airbyte import AirbyteConnection, AirbyteSyncMode

cereals_connection = AirbyteConnection(
    name="download-cereals",
    source=cereals_csv_source,
    destination=local_json_destination,
    stream_config={"cereals": AirbyteSyncMode.FULL_REFRESH_OVERWRITE},
)
```

We'll supply our new connection to the reconciler we defined earlier:

```python startafter=start_new_reconciler endbefore=end_new_reconciler file=/guides/dagster/managed_elements/airbyte.py dedent=4
airbyte_reconciler = AirbyteManagedElementReconciler(
    airbyte=airbyte_instance,
    connections=[cereals_connection],
)
```

## 4. Validating changes

> Next, we'll use the `dagster-me` CLI to sanity check our reconciler and apply any changes. The `check` command will print out any differences between the current state of the Airbyte instance and the desired state specified in the reconciler. The CLI can be invoked by pointing it at a module which contains the reconciler.

Reviewer: ```suggestion
Next, we'll use the `dagster-me` CLI to sanity-check our reconciler and apply any changes. The `check` command prints out differences between the current state of the Airbyte instance and the desired state specified in the reconciler. The CLI can be invoked by pointing it at a module that contains the reconciler.
```

A few small grammar tweaks here. Also suggesting active voice to improve clarity / readability.

```bash
dagster-me check --module my_python_module.my_submodule:reconciler

Found 1 managed elements, checking...
+ cereals-csv:
  + url: https://docs.dagster.io/assets/cereal.csv
  + format: csv
  + dataset_name: cereals
  + provider:
    + user_agent: False
    + storage: HTTPS
+ local-json:
  + destination_path: /local/cereals_out.json
+ download-cereals:
  + source: cereals-csv
  + destination: local-json
  + normalize data: None
  + streams:
    + cereals: FULL_REFRESH_OVERWRITE
```

## 5. Applying changes

These changes look like what we expect, so we'll apply them:

```bash
dagster-me apply --module my_python_module.my_submodule:reconciler
```

> Now, we should see our new connection in the Airbyte UI.

Reviewer: It'd be cool to include a screenshot of the connection in Airbyte here.

## 6. Loading connections into Dagster

> In order to load managed connections into Dagster, you can use the `load_assets_from_connections` utility method. This functions similarly to [`load_assets_from_airbyte_instance`](/integrations/airbyte#loading-airbyte-asset-definitions-from-an-airbyte-instance), but will validate that the connections you pass in match the connections defined in your Airbyte instance.

Reviewer: ```suggestion
In order to load managed connections into Dagster, you can use the `load_assets_from_connections` utility method. This functions similarly to [`load_assets_from_airbyte_instance`](/integrations/airbyte#loading-airbyte-asset-definitions-from-an-airbyte-instance), but validates that the connections you pass in match the connections defined in your Airbyte instance.
```

Suggesting active voice.
